crates/sui-adapter/src/programmable_transactions/context.rs

```diff
@@ -379,14 +654,82 @@ fn load_object_arg<S: Storage>(
     state_view: &S,
     object_owner_map: &mut BTreeMap<ObjectID, Owner>,
     obj_arg: ObjectArg,
-) -> Result<ObjectValue, ExecutionError> {
+) -> Result<InputValue, ExecutionError> {
     match obj_arg {
         ObjectArg::ImmOrOwnedObject((id, _, _))
         | ObjectArg::SharedObject {
             id, mutable: true, ..
-        } => load_object(state_view, object_owner_map, None, id),
+        } => load_object(
+            state_view,
+            object_owner_map,
+            /* imm override */ false,
+            id,
+        ),
         ObjectArg::SharedObject {
             id, mutable: false, ..
-        } => load_object(state_view, object_owner_map, Some(Owner::Immutable), id),
+        } => load_object(
+            state_view,
+            object_owner_map,
+            /* imm override */ true,
+            id,
+        ),
```

Comment:
nit: this might read a bit more clearly if you split it between shared and immutable

```
ObjectArg::ImmOrOwnedObject((id, _, _)) => load_object(
    state_view, 
    object_owner_map, 
    /* imm override */ false, 
    id,
),
ObjectArg::SharedObject { id, mutable, .. } => load_object(
    state_view, 
    object_owner_map, 
    !mutable, 
    id,
),
```